std::unix::fs::link using direct linkat call for Solaris.

devnexen opened the pull request:
Since we support solaris 11 as minimum, we can get rid of the runtime overhead.

try-job: dist-various-2

ChrisDenton commented on library/std/src/sys/pal/unix/fs.rs:
```diff
@@ -1753,8 +1753,8 @@ pub fn link(original: &Path, link: &Path) -> io::Result<()> {
-                } else if #[cfg(any(target_os = "macos", target_os = "solaris"))] {
-                    // MacOS (<=10.9) and Solaris 10 lack support for linkat while newer
+                } else if #[cfg(any(target_os = "macos"))] {
+                    // MacOS (<=10.9) lacks support for linkat while newer
```
It looks like our [minimum supported](https://doc.rust-lang.org/nightly/rustc/platform-support.html) macOS version is 10.12 so this branch can be removed entirely.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9489135279/job/26149749634) [(plain)](https://github.com/rust-lang/rust/commit/142e5b8491d05a109a41fa59317828abed6a9d0b/checks/26149749634/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

devnexen committed:
std::unix::fs::link using direct linkat call for Solaris and macOs.

Since we support solaris 11 and macOs Sierra as minimum, we can get rid
of the runtime overhead.

ChrisDenton commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 075f0686ad88b5f292b6a784406d668a3438b3ad has been approved by `ChrisDenton`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ChrisDenton 075f0686ad88b5f292b6a784406d668a3438b3ad -->
<!-- homu: {"type":"Approved","sha":"075f0686ad88b5f292b6a784406d668a3438b3ad","approver":"ChrisDenton","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
failed in rollup with dist-various-2:
```
   error: unused import: `libc::c_char`
    --> library/std/src/sys/pal/unix/fs.rs:35:5
     |
  35 | use libc::c_char;
     |     ^^^^^^^^^^^^
     |
     = note: `-D unused-imports` implied by `-D warnings`
     = help: to override `-D warnings` add `#[allow(unused_imports)]`
```

workingjubilee commented:
@bors r-

workingjubilee commented:
Please request a bors-try attempt after this patch is fixed and before it is r+'d again.

devnexen commented:
@bors try

bors commented:
@devnexen: :key: Insufficient privileges: not in try users

workingjubilee commented:
tfw you can't assign people permissions piecemeal.

@bors try

bors commented:
:hourglass: Trying commit c81ffab3ec66342d542ef7de39e1ec081027f43d with merge b7d907c7cce3c3019ae07d0b5128a133714dfd9a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c81ffab3ec66342d542ef7de39e1ec081027f43d","merge_sha":"b7d907c7cce3c3019ae07d0b5128a133714dfd9a"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9493717974/job/26164124436)
Build commit: b7d907c7cce3c3019ae07d0b5128a133714dfd9a (`b7d907c7cce3c3019ae07d0b5128a133714dfd9a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9493717974/job/26164124436"},"merge_sha":"b7d907c7cce3c3019ae07d0b5128a133714dfd9a"} -->

workingjubilee commented:
nice.

@bors r=ChrisDenton

bors commented:
:pushpin: Commit c81ffab3ec66342d542ef7de39e1ec081027f43d has been approved by `ChrisDenton`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ChrisDenton c81ffab3ec66342d542ef7de39e1ec081027f43d -->
<!-- homu: {"type":"Approved","sha":"c81ffab3ec66342d542ef7de39e1ec081027f43d","approver":"ChrisDenton","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
@bors rollup=

workingjubilee commented:
@bors rollup-
